### PR TITLE
Add the ability to specify the encryption algorithm

### PR DESCRIPTION
**Pull request by jdslavin**
- I use this for our project it to change the SHA value to "PBKDF2WithHmacSHA256" instead of the default of "PBKDF2WithHmacSHA1"

### CODE REVIEW

**Comment by weavejester** [comment]
This seems like a reasonable addition, but before I can merge you'll need to clean it up a little:

1. Remove any changes to the project version
2. Remove any changes to the profiles
3. Ensure the changed file has an ending newline
5. Remove the beginning "-" from the commit message
6. Use the imperative mood for the commit message, i.e. "Add" instead of "Added"
7. Remove the extra lines from the commit message, as it doesn't add anything to the explanation

You'll also need to:

1. Make a note of the default algorithm in the docstring
2. Add a simple test for the new functionality

**Comment by jdslavin** [comment]
I made the changes you requested, If I missed anything or you need something else let me know.

Thanks,

Jeff

**Comment by weavejester** [comment]
Your test is failing on the build server because the encryption algorithm you use for the tests doesn't exist. Try using the `SHA1` encryption algorithm instead.

You also repeat the previous test, and only test negative results. Instead, just test one successful and one failing test.

Can also you ensure that the lines you add are under 80 characters?

**Comment by jdslavin** [comment]
Sorry about that, i am using PBKDF2WithHmacSHA256 which only came available with Java JDK 1.8.  I changed the tests to use PBKDF2WithHmacSHA1 as I was not sure what would actually work on the integration server.

**Comment by weavejester** [comment]
Thanks for the update. The tests are passing on the CI server now. You now have two commits instead of one, though, and the indentation in the test files is now messed up. 

As I said in my previous post, we really only need to test one success and one failure case, so the tests I was thinking of were more something like:

```clojure
(let [salt      (byte-array (map byte [0 1 2 3 4 5 6 7]))
      algorithm "PBKDF2WithHmacSHA1"
      encrypt   #(password/encrypt % 100000 salt algorithm)
      check     #(password/check % algorithm)]
  (is (check (encrypt "foo") "foo"))
  (is (not (check (encrypt "foo") "bar"))))
```

However, now I write the tests out, I see that there's an issue with `password/check` - it shouldn't require more than one argument. The algorithm should be denoted in the encrypted string, but in such a way that it remains backward compatible.

I'm currently thinking of using:

    <iterations>$<salt>$<encrypted password>

For HMAC-SHA1 and:

    <iterations>$<salt>$<algorithm>$<encrypted password>

For all other algorithms. I think we should also convert "PBKDF2WithHmacSHA256" to simply "HMAC-SHA256", so we'd need a mapping like:

```clojure
(def algorithm-codes
  {"PBKDF2WithHmacSHA1" nil
   "PBKDF2WithHmacSHA256" "HMAC-SHA1"})
```

**Comment by jdslavin** [comment]
I like that idea of keeping the algorithm with the rest of the stored info so the check only needs the stored information, this would make it easier to change your algorithm out in the future for sure.    Do you want me to make those changes?

Jeff

**Comment by weavejester** [comment]
Go ahead :)

**Comment by jdslavin** [comment]
I got back to this today, after thinking about the backward compatibility issues with potentially saved passwords in the check function I added some logic and a test to account for previously stored passwords using this library.    

Also, i hope i got the commits in a way that is just a single commit.   

Thanks for all your feedback and help,

Jeff

File: src/crypto/password/pbkdf2.clj [hunk 1]
```diff
@@ -30,12 +38,15 @@
...
+     (encrypt raw iterations salt "HMAC-SHA1"))
+  ([raw iterations salt algorithm]
      (let [key-length  160
            key-spec    (PBEKeySpec. (.toCharArray raw) salt iterations key-length)
-           key-factory (SecretKeyFactory/getInstance "PBKDF2WithHmacSHA1")]
+           key-factory (SecretKeyFactory/getInstance (code-to-algorithm algorithm))]
        (->> (.generateSecret key-factory key-spec)
             (.getEncoded)
             (encode-str)
+            (str (encode-str (.getBytes algorithm)) "$")
```
Comment (weavejester) [inline src/crypto/password/pbkdf2.clj]: This could just be `(str algorithm "$")`. There's no need to encode the string, because we know the strings have no dangerous bytes.

File: src/crypto/password/pbkdf2.clj [hunk 2]
```diff
@@ -30,12 +38,15 @@
   ([raw iterations]
      (encrypt raw iterations (random/bytes 8)))
   ([raw iterations salt]
+     (encrypt raw iterations salt "HMAC-SHA1"))
+  ([raw iterations salt algorithm]
      (let [key-length  160
            key-spec    (PBEKeySpec. (.toCharArray raw) salt iterations key-length)
-           key-factory (SecretKeyFactory/getInstance "PBKDF2WithHmacSHA1")]
+           key-factory (SecretKeyFactory/getInstance (code-to-algorithm algorithm))]
```
Comment (weavejester) [inline src/crypto/password/pbkdf2.clj]: There should be a precondition to check there is an algorithm.

File: src/crypto/password/pbkdf2.clj [hunk 3]
```diff
@@ -9,6 +9,13 @@
            javax.crypto.spec.PBEKeySpec
            org.apache.commons.codec.binary.Base64))
 
+(def algorithm-codes
+  {"HMAC-SHA1" "PBKDF2WithHmacSHA1"
+   "HMAC-SHA256" "PBKDF2WithHmacSHA256"})
+
+(defn- code-to-algorithm [code]
```
Comment (weavejester) [inline src/crypto/password/pbkdf2.clj]: This function isn't necessary, since it's functionally equivalent to `algorithm-codes`.

File: src/crypto/password/pbkdf2.clj [hunk 4]
```diff
@@ -50,7 +61,13 @@
...
   otherwise."
   [raw encrypted]
-  (let [[i s _]    (str/split encrypted #"\$")
-        salt       (decode-str s)
-        iterations (decode-int i)]
-    (crypto/eq? encrypted (encrypt raw iterations salt))))
+   (let [[i s a h] (str/split encrypted #"\$")
+         algorithm (if (nil? h) "HMAC-SHA1" (apply str (map char (decode-str a))))
+         ; passwords in earlier versions of this library contained 3 parts,
+         ; need to convert to 4 part format for comparison
```
Comment (weavejester) [inline src/crypto/password/pbkdf2.clj]: I don't think we should do this. As I mentioned in my previous comment, I think that we should still produce a 3-part output for HMAC-SHA1, and 4-port for everything else. I don't think there's a compelling reason to do otherwise aside from aesthetics.

File: test/crypto/password/test/pbkdf2.clj [hunk 5]
```diff
@@ -1,20 +1,37 @@
...
+        algorithm "HMAC-SHA1"
+        encrypt #(password/encrypt % 100000 salt algorithm)
+        check #(password/check %1 %2)]
+    (is (check "foo" (encrypt "foo")))
+    (is (not (check "bar" (encrypt "foo")))))
+
+  ; compatibilty test for older 3 part passwords
+  (let [salt (byte-array (map byte [0 1 2 3 4 5 6 7]))
+        encrypt #(password/encrypt % 100000 salt)
+        remove-algorithm (fn [p] (let [[i s _ h] (str/split p #"\$")]
```
Comment (weavejester) [inline test/crypto/password/test/pbkdf2.clj]: This could be a separate `defn-` I think.

File: test/crypto/password/test/pbkdf2.clj [hunk 6]
```diff
@@ -1,20 +1,37 @@
...
-    "Some rather long pass phrase perhaps out of a book or poem")
-  
+           "a"
+           "foo"
+           "password"
+           "Testing"
+           "Test123"
+           "ÁäñßOÔ"
+           "großpösna"
+           "Some rather long pass phrase perhaps out of a book or poem")
```
Comment (weavejester) [inline test/crypto/password/test/pbkdf2.clj]: You need to fix the indentation that your commit changed here.

File: test/crypto/password/test/pbkdf2.clj [hunk 7]
```diff
@@ -1,20 +1,37 @@
...
+
   (are [s r] (not (password/check r (password/encrypt s)))
-    "a" "b"
-    "a" "a "
-    "aaaaa" "aaaaa\n"
-    "großpösna" "grossposna"))
+             "a" "b"
```
Comment (weavejester) [inline test/crypto/password/test/pbkdf2.clj]: And the indentation here.

**Comment by weavejester** [comment]
Sorry for taking a while to get back to you on this. There are a few minor formatting and other issues, but nothing that affects the operation of the code, so I'll merge and sort them out in a separate commit.
